File: apps/mobile/modules/clawket-voice-capture/android/src/main/java/com/clawket/voicecapture/ClawketVoiceCaptureModule.kt
package com.clawket.voicecapture

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.PackageManager
import android.media.AudioDeviceInfo
import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioRecord
import android.media.MediaRecorder
import android.os.Process
import android.os.SystemClock
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.jni.NativeArrayBuffer
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.lang.ref.WeakReference
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.concurrent.Executors

private const val BUFFER_EVENT = "onVoiceCaptureBuffer"
private const val STATUS_EVENT = "onVoiceCaptureStatus"
private const val CHUNK_MS = 40
private val SAMPLE_RATES = intArrayOf(16000, 48000, 44100)

internal class VoiceCapturePermissionException :
  CodedException("RECORD_AUDIO permission has not been granted")

internal class VoiceCaptureStartException(message: String) :
  CodedException(message)

class ClawketVoiceCaptureModule : Module() {
  internal val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  override fun definition() = ModuleDefinition {
    Name("ClawketVoiceCapture")

    Events(BUFFER_EVENT, STATUS_EVENT)

    OnCreate {
      VoiceCaptureController.attach(this@ClawketVoiceCaptureModule)
    }

    OnDestroy {
      VoiceCaptureController.detach(this@ClawketVoiceCaptureModule)
    }

    AsyncFunction("prepare") { promise: Promise ->
      VoiceCaptureController.run(promise) { prepare(this@ClawketVoiceCaptureModule.context) }
    }

    AsyncFunction("start") { captureId: String, requestedAt: Double, promise: Promise ->
      VoiceCaptureController.run(promise) { start(this@ClawketVoiceCaptureModule.context, captureId, requestedAt) }
    }

    AsyncFunction("stop") { captureId: String, promise: Promise ->
      VoiceCaptureController.run(promise) { stop(captureId) }
    }

    AsyncFunction("release") { promise: Promise ->
      VoiceCaptureController.run(promise) { release() }
    }
  }

  internal fun emit(event: String, body: Map<String, Any?>) {
    sendEvent(event, body)
  }
}

/**
 * One process-wide recorder. The AudioRecord is created before the first press and reused;
 * nothing is recorded until `start`. Control work runs on a dedicated thread instead of the
 * shared Expo module queue; samples are read on their own urgent-audio thread in 40 ms chunks.
 */
internal object VoiceCaptureController {
  private val control = Executors.newSingleThreadExecutor { runnable ->
    Thread({
      Process.setThreadPriority(Process.THREAD_PRIORITY_URGENT_AUDIO)
      runnable.run()
    }, "clawket-voice-control")
  }

  @Volatile private var module: WeakReference<ClawketVoiceCaptureModule>? = null
  @Volatile private var captureId: String? = null
  private var recorder: AudioRecord? = null
  private var reader: Thread? = null
  private var releasePending = false

  fun run(promise: Promise, block: VoiceCaptureController.() -> Any?) {
    control.execute {
      try {
        val result = block()
        promise.resolve(if (result == Unit) null else result)
      } catch (error: CodedException) {
        promise.reject(error)
      } catch (error: Throwable) {
        promise.reject(VoiceCaptureStartException(error.message ?: "The microphone could not start"))
      }
    }
  }

  fun attach(owner: ClawketVoiceCaptureModule) {
    module = WeakReference(owner)
  }

  fun detach(owner: ClawketVoiceCaptureModule) {
    control.execute {
      val current = module?.get()
      if (current != null && current !== owner) return@execute
      module = null
      captureId?.let { finishCapture() }
      teardown()
    }
  }

  // Control thread only.

  fun prepare(context: Context) {
    releasePending = false
    if (captureId != null || recorder != null || !granted(context)) return
    recorder = createRecorder()
  }

  fun start(context: Context, id: String, requestedAt: Double): Map<String, Any> {
    val queueMs = (System.currentTimeMillis() - requestedAt).coerceAtLeast(0.0)
    if (!granted(context)) throw VoiceCapturePermissionException()
    if (captureId != null) finishCapture()
    releasePending = false

    val warm = recorder != null
    val preparation = SystemClock.elapsedRealtime()
    val record = recorder ?: createRecorder().also { recorder = it }
    val engineMs = (SystemClock.elapsedRealtime() - preparation).toDouble()

    val launch = SystemClock.elapsedRealtime()
    try {
      record.startRecording()
    } catch (error: IllegalStateException) {
      discardRecorder()
      throw VoiceCaptureStartException("The microphone is held by another app or the system")
    }
    if (record.recordingState != AudioRecord.RECORDSTATE_RECORDING) {
      discardRecorder()
      throw VoiceCaptureStartException("The microphone is held by another app or the system")
    }
    val startMs = (SystemClock.elapsedRealtime() - launch).toDouble()
    captureId = id
    startReader(record, id)

    val audio = context.getSystemService(Context.AUDIO_SERVICE) as? AudioManager
    return mapOf(
      "queueMs" to queueMs,
      "activateMs" to 0.0,
      "engineMs" to engineMs,
      "startMs" to startMs,
      "warm" to warm,
      "inputRoute" to inputRoute(record.routedDevice),
      "bluetooth" to bluetoothConnected(audio),
      "otherAudio" to (audio?.isMusicActive == true)
    )
  }

  fun stop(id: String) {
    if (captureId != id) return
    finishCapture()
    if (releasePending) teardown()
  }

  fun release() {
    if (captureId != null) {
      // The owning attempt's stop completes the release; another screen never stops it.
      releasePending = true
      return
    }
    teardown()
  }

  private fun finishCapture() {
    captureId = null
    try {
      recorder?.stop()
    } catch (_: IllegalStateException) {
      // Already stopped.
    }
    reader?.let { if (it !== Thread.currentThread()) it.join(500) }
    reader = null
  }

  private fun fail(id: String) {
    if (captureId != id) return
    finishCapture()
    discardRecorder()
    module?.get()?.emit(STATUS_EVENT, mapOf("captureId" to id, "reason" to "failed"))
  }

  private fun teardown() {
    releasePending = false
    discardRecorder()
  }

  private fun discardRecorder() {
    recorder?.release()
    recorder = null
  }

  @SuppressLint("MissingPermission")
  private fun createRecorder(): AudioRecord {
    val channel = AudioFormat.CHANNEL_IN_MONO
    val encoding = AudioFormat.ENCODING_PCM_FLOAT
    for (rate in SAMPLE_RATES) {
      val minimum = AudioRecord.getMinBufferSize(rate, channel, encoding)
      if (minimum <= 0) continue
      // At least 200 ms of headroom so the 40 ms reader never overruns.
      val bufferSize = maxOf(minimum, rate * 4 / 5)
      val record = try {
        AudioRecord(MediaRecorder.AudioSource.MIC, rate, channel, encoding, bufferSize)
      } catch (_: Exception) {
        null
      }
      if (record?.state == AudioRecord.STATE_INITIALIZED) return record
      record?.release()
    }
    throw VoiceCaptureStartException("The microphone is unavailable")
  }

  private fun startReader(record: AudioRecord, id: String) {
    val frames = record.sampleRate * CHUNK_MS / 1000
    val rate = record.sampleRate.toDouble()
    val thread = Thread({
      Process.setThreadPriority(Process.THREAD_PRIORITY_URGENT_AUDIO)
      val samples = FloatArray(frames)
      while (captureId == id) {
        val read = record.read(samples, 0, frames, AudioRecord.READ_BLOCKING)
        if (captureId != id) break
        if (read > 0) {
          val bytes = ByteBuffer.allocateDirect(read * 4).order(ByteOrder.nativeOrder())
          bytes.asFloatBuffer().put(samples, 0, read)
          module?.get()?.emit(BUFFER_EVENT, mapOf(
            "captureId" to id,
            "data" to NativeArrayBuffer(bytes),
            "sampleRate" to rate,
            "channels" to 1
          ))
        } else if (read < 0) {
          control.execute { fail(id) }
          break
        }
      }
    }, "clawket-voice-reader")
    reader = thread
    thread.start()
  }

  private fun granted(context: Context): Boolean =
    context.checkSelfPermission(Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED

  /** Device categories only; product names are never reported. */
  private fun inputRoute(device: AudioDeviceInfo?): String = when (device?.type) {
    null -> "none"
    AudioDeviceInfo.TYPE_BUILTIN_MIC -> "built_in"
    AudioDeviceInfo.TYPE_BLUETOOTH_SCO, AudioDeviceInfo.TYPE_BLE_HEADSET -> "bluetooth"
    AudioDeviceInfo.TYPE_WIRED_HEADSET -> "wired"
    AudioDeviceInfo.TYPE_USB_DEVICE, AudioDeviceInfo.TYPE_USB_HEADSET -> "usb"
    else -> "other"
  }

  private fun bluetoothConnected(audio: AudioManager?): Boolean {
    val devices = audio?.getDevices(AudioManager.GET_DEVICES_OUTPUTS) ?: return false
    return devices.any {
      it.type == AudioDeviceInfo.TYPE_BLUETOOTH_A2DP || it.type == AudioDeviceInfo.TYPE_BLUETOOTH_SCO ||
        it.type == AudioDeviceInfo.TYPE_BLE_HEADSET
    }
  }
}

File: apps/mobile/modules/clawket-voice-capture/ios/ClawketVoiceCaptureModule.swift
import AVFoundation
import ExpoModulesCore

/// Every audio-session and engine transition runs here, never on Expo's shared async queue,
/// so Keychain, file and analytics work cannot delay the microphone (or be delayed by it).
private let voiceCaptureQueue = DispatchQueue(label: "com.clawket.voice-capture", qos: .userInteractive)

public final class ClawketVoiceCaptureModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ClawketVoiceCapture")

    Events(VoiceCaptureController.bufferEvent, VoiceCaptureController.statusEvent)

    OnCreate {
      VoiceCaptureController.shared.attach(self)
    }

    OnDestroy {
      VoiceCaptureController.shared.detach(self)
    }

    OnAppEntersBackground {
      VoiceCaptureController.shared.releaseWhenIdle()
    }

    AsyncFunction("prepare") {
      try VoiceCaptureController.shared.prepare()
    }.runOnQueue(voiceCaptureQueue)

    AsyncFunction("start") { (captureId: String, requestedAt: Double) -> [String: Any] in
      try VoiceCaptureController.shared.start(captureId: captureId, requestedAt: requestedAt)
    }.runOnQueue(voiceCaptureQueue)

    AsyncFunction("stop") { (captureId: String) in
      VoiceCaptureController.shared.stop(captureId: captureId)
    }.runOnQueue(voiceCaptureQueue)

    AsyncFunction("release") {
      VoiceCaptureController.shared.release()
    }.runOnQueue(voiceCaptureQueue)
  }
}

/**
 One process-wide microphone. The session category is configured once, the engine and its tap
 stay allocated between recordings, and the session stays active while idle unless activating it
 interrupted another app's audio. Nothing is captured until `start`; the engine is paused, not
 running, between recordings. All mutable state is confined to `voiceCaptureQueue`, except the
 emit target that the tap thread reads under `emitLock`.
 */
final class VoiceCaptureController {
  static let shared = VoiceCaptureController()
  static let bufferEvent = "onVoiceCaptureBuffer"
  static let statusEvent = "onVoiceCaptureStatus"
  private static let targetSampleRate: Double = 16000
  private static let idleDeactivationSeconds: TimeInterval = 30

  private weak var module: ClawketVoiceCaptureModule?
  private var engine: AVAudioEngine?
  private var configurationObserver: NSObjectProtocol?
  private var tapFormat: AVAudioFormat?
  private var tapConverter: AVAudioConverter?
  private var sessionActive = false
  private var interruptedOthers = false
  private var captureId: String?
  private var releasePending = false
  private var idleDeactivation: DispatchWorkItem?

  private let emitLock = NSLock()
  private var emitTarget: (id: String, module: ClawketVoiceCaptureModule)?

  private init() {
    let center = NotificationCenter.default
    let session = AVAudioSession.sharedInstance()
    _ = center.addObserver(forName: AVAudioSession.interruptionNotification, object: session, queue: nil) { note in
      guard let raw = note.userInfo?[AVAudioSessionInterruptionTypeKey] as? UInt,
        AVAudioSession.InterruptionType(rawValue: raw) == .began else {
        return
      }
      voiceCaptureQueue.async { VoiceCaptureController.shared.handleInterruption() }
    }
    _ = center.addObserver(forName: AVAudioSession.mediaServicesWereResetNotification, object: session, queue: nil) { _ in
      voiceCaptureQueue.async { VoiceCaptureController.shared.handleMediaServicesReset() }
    }
  }

  // MARK: - Module lifecycle

  func attach(_ module: ClawketVoiceCaptureModule) {
    voiceCaptureQueue.async { self.module = module }
  }

  func detach(_ module: ClawketVoiceCaptureModule) {
    voiceCaptureQueue.async {
      guard self.module == nil || self.module === module else {
        return
      }
      self.module = nil
      if let id = self.captureId {
        self.finishCapture(id)
      }
      self.teardown()
    }
  }

  func releaseWhenIdle() {
    voiceCaptureQueue.async { self.release() }
  }

  // MARK: - Commands (voiceCaptureQueue)

  /// Warms everything that does not open the microphone or interrupt other audio.
  func prepare() throws {
    releasePending = false
    guard captureId == nil, recordPermissionGranted() else {
      return
    }
    try configureCategory()
    let engine = ensureEngine()
    let format = engine.inputNode.outputFormat(forBus: 0)
    if sessionActive, isUsable(format) {
      if tapFormat != format {
        installTap(on: engine, format: format)
      }
      engine.prepare()
    }
  }

  func start(captureId id: String, requestedAt: Double) throws -> [String: Any] {
    let queueMs = max(0, Date().timeIntervalSince1970 * 1000 - requestedAt)
    guard recordPermissionGranted() else {
      throw VoiceCapturePermissionException()
    }
    if let current = captureId {
      finishCapture(current)
    }
    releasePending = false
    cancelIdleDeactivation()

    let session = AVAudioSession.sharedInstance()
    let warm = sessionActive && engine != nil && tapFormat != nil
    let bluetooth = routeUsesBluetooth(session.currentRoute)
    var otherAudio = false
    var activateMs = 0.0
    do {
      try configureCategory()
      if !sessionActive {
        otherAudio = session.isOtherAudioPlaying
        let activation = Date()
        try session.setActive(true)
        activateMs = milliseconds(since: activation)
        sessionActive = true
        interruptedOthers = otherAudio
      }

      let preparation = Date()
      var engine = ensureEngine()
      var format = engine.inputNode.outputFormat(forBus: 0)
      guard isUsable(format) else {
        throw VoiceCaptureUnavailableException()
      }
      if tapFormat != format {
        installTap(on: engine, format: format)
      }
      tapConverter?.reset()
      let engineMs = milliseconds(since: preparation)

      setEmitTarget(id)
      let launch = Date()
      do {
        try engine.start()
      } catch {
        // A paused engine can refuse to resume after a route change; rebuild it once.
        teardownEngine()
        engine = ensureEngine()
        format = engine.inputNode.outputFormat(forBus: 0)
        guard isUsable(format) else {
          throw VoiceCaptureUnavailableException()
        }
        installTap(on: engine, format: format)
        try engine.start()
      }
      let startMs = milliseconds(since: launch)
      captureId = id
      return [
        "queueMs": queueMs,
        "activateMs": activateMs,
        "engineMs": engineMs,
        "startMs": startMs,
        "warm": warm,
        "inputRoute": inputRoute(session.currentRoute),
        "bluetooth": bluetooth,
        "otherAudio": otherAudio
      ]
    } catch {
      clearEmitTarget()
      engine?.pause()
      settleIdleSession()
      if error is Exception {
        throw error
      }
      throw VoiceCaptureStartException(error.localizedDescription)
    }
  }

  func stop(captureId id: String) {
    guard captureId == id else {
      return
    }
    finishCapture(id)
    settleIdleSession()
  }

  func release() {
    cancelIdleDeactivation()
    if captureId != nil {
      // The owning attempt's stop completes the release; another screen never stops it.
      releasePending = true
      return
    }
    teardown()
  }

  // MARK: - Session and engine

  private func configureCategory() throws {
    let session = AVAudioSession.sharedInstance()
    let options: AVAudioSession.CategoryOptions = [.defaultToSpeaker]
    if session.category != .playAndRecord || session.mode != .measurement || session.categoryOptions != options {
      try session.setCategory(.playAndRecord, mode: .measurement, options: options)
    }
    // Start/stop haptics and slide-to-cancel feedback must stay perceptible while recording.
    if !session.allowHapticsAndSystemSoundsDuringRecording {
      try? session.setAllowHapticsAndSystemSoundsDuringRecording(true)
    }
  }

  private func ensureEngine() -> AVAudioEngine {
    if let engine {
      return engine
    }
    let engine = AVAudioEngine()
    // Instantiates the input unit now instead of on the press.
    _ = engine.inputNode
    configurationObserver = NotificationCenter.default.addObserver(
      forName: .AVAudioEngineConfigurationChange,
      object: engine,
      queue: nil
    ) { [weak engine] _ in
      voiceCaptureQueue.async { VoiceCaptureController.shared.handleConfigurationChange(engine) }
    }
    self.engine = engine
    return engine
  }

  private func installTap(on engine: AVAudioEngine, format: AVAudioFormat) {
    let input = engine.inputNode
    if tapFormat != nil {
      input.removeTap(onBus: 0)
    }
    guard let target = AVAudioFormat(
      commonFormat: .pcmFormatFloat32,
      sampleRate: Self.targetSampleRate,
      channels: 1,
      interleaved: true
    ) else {
      tapFormat = nil
      return
    }
    // The hardware rate is never forced to 16 kHz; conversion happens here instead.
    let matchesTarget = format.sampleRate == Self.targetSampleRate && format.channelCount == 1
      && format.commonFormat == .pcmFormatFloat32
    let converter = matchesTarget ? nil : AVAudioConverter(from: format, to: target)
    converter?.downmix = true
    tapConverter = converter
    let bufferSize = AVAudioFrameCount(max(1, format.sampleRate * 0.1))
    input.installTap(onBus: 0, bufferSize: bufferSize, format: format) { buffer, _ in
      VoiceCaptureController.shared.deliver(buffer, converter: converter, target: target)
    }
    tapFormat = format
  }

  private func finishCapture(_ id: String) {
    guard captureId == id || captureId == nil else {
      return
    }
    clearEmitTarget()
    captureId = nil
    // Pausing stops the hardware (and the microphone indicator) but keeps the engine allocated.
    engine?.pause()
  }

  private func settleIdleSession() {
    if releasePending {
      teardown()
    } else if interruptedOthers {
      // Let the interrupted app resume now, exactly as a released recorder would.
      deactivate()
    } else {
      scheduleIdleDeactivation()
    }
  }

  private func teardown() {
    releasePending = false
    teardownEngine()
    deactivate()
  }

  private func teardownEngine() {
    if let configurationObserver {
      NotificationCenter.default.removeObserver(configurationObserver)
    }
    configurationObserver = nil
    if let engine {
      if tapFormat != nil {
        engine.inputNode.removeTap(onBus: 0)
      }
      engine.stop()
    }
    engine = nil
    tapFormat = nil
    tapConverter = nil
  }

  private func deactivate() {
    cancelIdleDeactivation()
    guard sessionActive else {
      return
    }
    engine?.stop()
    try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
    sessionActive = false
    interruptedOthers = false
  }

  private func scheduleIdleDeactivation() {
    cancelIdleDeactivation()
    let work = DispatchWorkItem {
      let controller = VoiceCaptureController.shared
      if controller.captureId == nil {
        controller.deactivate()
      }
    }
    idleDeactivation = work
    voiceCaptureQueue.asyncAfter(deadline: .now() + Self.idleDeactivationSeconds, execute: work)
  }

  private func cancelIdleDeactivation() {
    idleDeactivation?.cancel()
    idleDeactivation = nil
  }

  // MARK: - System notifications (voiceCaptureQueue)

  private func handleInterruption() {
    cancelIdleDeactivation()
    sessionActive = false
    interruptedOthers = false
    guard let id = captureId else {
      return
    }
    let module = self.module
    finishCapture(id)
    module?.emit(event: Self.statusEvent, payload: ["captureId": id, "reason": "interrupted"])
  }

  private func handleConfigurationChange(_ changed: AVAudioEngine?) {
    guard let engine, let changed, engine === changed else {
      return
    }
    // The engine stopped itself because the route or hardware format changed.
    let format = engine.inputNode.outputFormat(forBus: 0)
    guard let id = captureId else {
      if tapFormat != nil {
        engine.inputNode.removeTap(onBus: 0)
      }
      tapFormat = nil
      tapConverter = nil
      return
    }
    if isUsable(format) {
      if tapFormat != format {
        installTap(on: engine, format: format)
      }
      if (try? engine.start()) != nil {
        return
      }
    }
    let module = self.module
    finishCapture(id)
    module?.emit(event: Self.statusEvent, payload: ["captureId": id, "reason": "route"])
  }

  private func handleMediaServicesReset() {
    let id = captureId
    let module = self.module
    clearEmitTarget()
    captureId = nil
    if let configurationObserver {
      NotificationCenter.default.removeObserver(configurationObserver)
    }
    // The old engine and session state are invalid after a reset; rebuild on the next press.
    configurationObserver = nil
    engine = nil
    tapFormat = nil
    tapConverter = nil
    sessionActive = false
    interruptedOthers = false
    cancelIdleDeactivation()
    if let id {
      module?.emit(event: Self.statusEvent, payload: ["captureId": id, "reason": "reset"])
    }
  }

  // MARK: - Tap thread

  private func deliver(_ buffer: AVAudioPCMBuffer, converter: AVAudioConverter?, target: AVAudioFormat) {
    guard let emitter = currentEmitTarget(), buffer.frameLength > 0 else {
      return
    }
    var output = buffer
    if let converter {
      let capacity = AVAudioFrameCount(Double(buffer.frameLength) * target.sampleRate / buffer.format.sampleRate) + 1
      guard let converted = AVAudioPCMBuffer(pcmFormat: target, frameCapacity: capacity) else {
        return
      }
      var consumed = false
      var error: NSError?
      converter.convert(to: converted, error: &error) { _, status in
        if consumed {
          status.pointee = .noDataNow
          return nil
        }
        consumed = true
        status.pointee = .haveData
        return buffer
      }
      guard error == nil, converted.frameLength > 0 else {
        return
      }
      output = converted
    }
    // Mono float samples; a multi-channel buffer without a converter contributes its first channel.
    guard let samples = output.floatChannelData?[0] else {
      return
    }
    let data = NativeArrayBuffer.copy(of: samples, count: Int(output.frameLength) * MemoryLayout<Float>.size)
    let payload: [String: Any] = [
      "captureId": emitter.id,
      "data": data,
      "sampleRate": output.format.sampleRate,
      "channels": 1
    ]
    emitter.module.emit(event: Self.bufferEvent, payload: payload)
  }

  private func setEmitTarget(_ id: String) {
    emitLock.lock()
    defer { emitLock.unlock() }
    emitTarget = module.map { (id: id, module: $0) }
  }

  private func clearEmitTarget() {
    emitLock.lock()
    defer { emitLock.unlock() }
    emitTarget = nil
  }

  private func currentEmitTarget() -> (id: String, module: ClawketVoiceCaptureModule)? {
    emitLock.lock()
    defer { emitLock.unlock() }
    return emitTarget
  }

  // MARK: - Helpers

  private func isUsable(_ format: AVAudioFormat) -> Bool {
    format.sampleRate > 0 && format.channelCount > 0
  }

  private func recordPermissionGranted() -> Bool {
    if #available(iOS 17.0, *) {
      return AVAudioApplication.shared.recordPermission == .granted
    }
    return AVAudioSession.sharedInstance().recordPermission == .granted
  }

  private func milliseconds(since date: Date) -> Double {
    (Date().timeIntervalSince(date) * 1000).rounded()
  }

  /// Port categories only; device names are never reported.
  private func inputRoute(_ route: AVAudioSessionRouteDescription) -> String {
    guard let port = route.inputs.first?.portType else {
      return "none"
    }
    switch port {
    case .builtInMic:
      return "built_in"
    case .bluetoothHFP, .bluetoothLE:
      return "bluetooth"
    case .headsetMic:
      return "wired"
    case .usbAudio:
      return "usb"
    case .carAudio:
      return "car"
    default:
      return "other"
    }
  }

  private func routeUsesBluetooth(_ route: AVAudioSessionRouteDescription) -> Bool {
    let bluetooth: Set<AVAudioSession.Port> = [.bluetoothA2DP, .bluetoothHFP, .bluetoothLE]
    return (route.inputs + route.outputs).contains { bluetooth.contains($0.portType) }
  }
}

internal final class VoiceCapturePermissionException: Exception, @unchecked Sendable {
  override var reason: String {
    "Microphone permission has not been granted"
  }
}

internal final class VoiceCaptureUnavailableException: Exception, @unchecked Sendable {
  override var reason: String {
    "The microphone input is unavailable"
  }
}

internal final class VoiceCaptureStartException: GenericException<String>, @unchecked Sendable {
  override var reason: String {
    "The microphone could not start: \(param)"
  }
}
